File: python/python-rest/src/com/intellij/python/reStructuredText/run/sphinx/SphinxRunConfiguration.java
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.python.reStructuredText.run.sphinx;

import com.intellij.execution.ExecutionException;
import com.intellij.execution.Executor;
import com.intellij.execution.configurations.*;
import com.intellij.execution.runners.ExecutionEnvironment;
import com.intellij.openapi.fileChooser.FileChooserDescriptorFactory;
import com.intellij.openapi.options.SettingsEditor;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.python.reStructuredText.run.RestConfigurationEditor;
import com.intellij.python.reStructuredText.run.RestRunConfiguration;
import com.jetbrains.python.PyPsiPackageUtil;
import com.jetbrains.python.packaging.PyPackage;
import com.jetbrains.python.packaging.PyPackageManager;
import com.intellij.python.reStructuredText.PythonRestBundle;
import org.jetbrains.annotations.NotNull;

import java.util.List;

/**
 * User : catherine
 */
public class SphinxRunConfiguration extends RestRunConfiguration {
  public SphinxRunConfiguration(final Project project,
                                final ConfigurationFactory factory) {
    super(project, factory);
  }

  @Override
  protected SettingsEditor<? extends RunConfiguration> createConfigurationEditor() {
    final SphinxTasksModel model = new SphinxTasksModel();
    if (!model.contains("pdf") && getSdk() != null) {
      final List<PyPackage> packages = PyPackageManager.getInstance(getSdk()).getPackages();
      if (packages != null) {
        final PyPackage rst2pdf = PyPsiPackageUtil.findPackage(packages, "rst2pdf");
        if (rst2pdf != null) {
          model.add(13, "pdf");
        }
      }
    }

    RestConfigurationEditor editor = new RestConfigurationEditor(getProject(), this, model);
    editor.setConfigurationName("Sphinx task");
    editor.setOpenInBrowserVisible(false);
    editor.setInputDescriptor(FileChooserDescriptorFactory.createSingleFolderDescriptor());
    editor.setOutputDescriptor(FileChooserDescriptorFactory.createSingleFolderDescriptor());
    return editor;
  }

  @Override
  public RunProfileState getState(@NotNull Executor executor, @NotNull ExecutionEnvironment env) throws ExecutionException {
    return new SphinxCommandLineState(this, env);
  }

  @Override
  public void checkConfiguration() throws RuntimeConfigurationException {
    super.checkConfiguration();
    if (StringUtil.isEmptyOrSpaces(getInputFile()))
      throw new RuntimeConfigurationError(PythonRestBundle.message("python.rest.specify.input.directory.name"));
    if (StringUtil.isEmptyOrSpaces(getOutputFile()))
      throw new RuntimeConfigurationError(PythonRestBundle.message("python.rest.specify.output.directory.name"));
  }

  @Override
  public String suggestedName() {
    return PythonRestBundle.message("python.rest.sphinx.run.cfg.default.name", getName());
  }
}

File: python/python-restructuredtext/src/com/intellij/restructuredtext/python/run/sphinx/SphinxRunConfiguration.kt
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.restructuredtext.python.run.sphinx

import com.intellij.execution.ExecutionException
import com.intellij.execution.Executor
import com.intellij.execution.configurations.*
import com.intellij.execution.runners.ExecutionEnvironment
import com.intellij.openapi.fileChooser.FileChooserDescriptorFactory
import com.intellij.openapi.options.SettingsEditor
import com.intellij.openapi.project.Project
import com.intellij.restructuredtext.python.PythonRestBundle.message
import com.intellij.restructuredtext.python.run.RestConfigurationEditor
import com.intellij.restructuredtext.python.run.RestRunConfiguration
import com.jetbrains.python.packaging.management.PythonPackageManager
import com.jetbrains.python.packaging.management.hasInstalledPackageSnapshot

/**
 * User : catherine
 */
class SphinxRunConfiguration(
  project: Project?,
  factory: ConfigurationFactory?,
) : RestRunConfiguration(project, factory) {
  override fun createConfigurationEditor(): SettingsEditor<out RunConfiguration?> {
    val model = SphinxTasksModel()
    val sdk = sdk
    if (!model.contains("pdf") && sdk != null) {
      val packageManager = PythonPackageManager.forSdk(project, sdk)
      val isInstalled = packageManager.hasInstalledPackageSnapshot("rst2pdf")
      if (isInstalled) {
        model.add(13, "pdf")
      }
    }

    val editor = RestConfigurationEditor(project, this, model)
    editor.setConfigurationName("Sphinx task")
    editor.setOpenInBrowserVisible(false)
    editor.setInputDescriptor(FileChooserDescriptorFactory.createSingleFolderDescriptor())
    editor.setOutputDescriptor(FileChooserDescriptorFactory.createSingleFolderDescriptor())
    return editor
  }

  @Throws(ExecutionException::class)
  override fun getState(executor: Executor, env: ExecutionEnvironment): RunProfileState {
    return SphinxCommandLineState(this, env)
  }

  @Throws(RuntimeConfigurationException::class)
  override fun checkConfiguration() {
    super.checkConfiguration()
    if (inputFile.isNullOrBlank()) throw RuntimeConfigurationError(message("python.rest.specify.input.directory.name"))
    if (outputFile.isNullOrBlank()) throw RuntimeConfigurationError(message("python.rest.specify.output.directory.name"))
  }

  override fun suggestedName(): String {
    return message("python.rest.sphinx.run.cfg.default.name", name)
  }
}
